### ISSUE TITLE
jfrog-cli rt dl appears to ignore/override umask

### ISSUE THREAD

**Issue by nneul** [jfrog/jfrog-cli-go#193]
Downloading an artifact with 'jfrog rt dl' is resulting in a 0600 file despite 022 umask being set. Appears to only occur for container type artifacts (.jar, .tar), but that may just be due to small sample size.

**Comment by scottrus** [comment]
Confirmed. This is not seen on all files but on tgz files I'm seeing the same. The problem was introduced in 1.16.0 for Linux 64. 


Here is what version 1.15.1 looks like on a simple download of a TGZ file:

```
[srussell@ln70-srussell jfrog_1.15.1]$ ./jfrog -v
jfrog version 1.15.1
[srussell@ln70-srussell jfrog_1.15.1]$ ./jfrog rt dl builds-local/third_party-dev_default/315/third_party-577247-default-9607b03522d0-linux.tgz
[Info] Searching items to download...
[Debug] Searching Artifactory using AQL query:
 items.find({"repo": "builds-local","path": {"$ne": "."},"$or": [{"$and":[{"path": {"$match": "third_party-dev_default/315"},"name": {"$match": "third_party-577247-default-9607b03522d0-linux.tgz"}}]}]}).include("name","repo","path","actual_md5","actual_sha1","size","type","property")
[Debug] Artifactory response:  200 OK
[Info] [Thread 2] Downloading builds-local/third_party-dev_default/315/third_party-577247-default-9607b03522d0-linux.tgz
[Info] [Thread 2] [0]: 206 Partial Content...
[Info] [Thread 2] [1]: 206 Partial Content...
[Info] [Thread 2] [2]: 206 Partial Content...
[Info] [Thread 2] Done downloading.
[Debug] Downloaded 1 artifacts.
{
  "status": "success",
  "totals": {
    "success": 1,
    "failure": 0
  }
}
[srussell@ln70-srussell jfrog_1.15.1]$ ls -alR
.:
total 19692
drwxr-xr-x  3 srussell eng     4096 Jul 26 14:26 .
drwxr-xr-x 68 srussell eng    12288 Jul 26 14:21 ..
-rwxr-xr-x  1 srussell eng 20141527 May  3 09:30 jfrog
drwxr-xr-x  3 srussell eng     4096 Jul 26 14:26 third_party-dev_default

./third_party-dev_default:
total 12
drwxr-xr-x 3 srussell eng 4096 Jul 26 14:26 .
drwxr-xr-x 3 srussell eng 4096 Jul 26 14:26 ..
drwxr-xr-x 2 srussell eng 4096 Jul 26 14:26 315

./third_party-dev_default/315:
total 124424
drwxr-xr-x 2 srussell eng      4096 Jul 26 14:26 .
drwxr-xr-x 3 srussell eng      4096 Jul 26 14:26 ..
-rw-r--r-- 1 srussell eng 127399504 Jul 26 14:26 third_party-577247-default-9607b03522d0-linux.tgz
```

Perms on the above file are 644. Now running the same command with 1.16.0:

```
[srussell@ln70-srussell jfrog_1.16.0]$ ./jfrog -v
jfrog version 1.16.0
[srussell@ln70-srussell jfrog_1.16.0]$ ./jfrog rt dl builds-local/third_party-dev_default/315/third_party-577247-default-9607b03522d0-linux.tgz
[Info] Searching items to download...
[Debug] Searching Artifactory using AQL query:
 items.find({"repo": "builds-local","path": {"$ne": "."},"$or": [{"$and":[{"path": {"$match": "third_party-dev_default/315"},"name": {"$match": "third_party-577247-default-9607b03522d0-linux.tgz"}}]}]}).include("name","repo","path","actual_md5","actual_sha1","size","type","property")
[Debug] Artifactory response:  200 OK
[Info] [Thread 2] Downloading builds-local/third_party-dev_default/315/third_party-577247-default-9607b03522d0-linux.tgz
[Info] [Thread 2] [2]: 206 Partial Content...
[Info] [Thread 2] [0]: 206 Partial Content...
[Info] [Thread 2] [1]: 206 Partial Content...
[Info] [Thread 2] Done downloading.
[Debug] Downloaded 1 artifacts.
{
  "status": "success",
  "totals": {
    "success": 1,
    "failure": 0
  }
}
[srussell@ln70-srussell jfrog_1.16.0]$ ls -alR
.:
total 19712
drwxr-xr-x  3 srussell eng     4096 Jul 26 14:25 .
drwxr-xr-x 68 srussell eng    12288 Jul 26 14:21 ..
-rwxr-xr-x  1 srussell eng 20163078 May 16 10:55 jfrog
drwxr-xr-x  3 srussell eng     4096 Jul 26 14:25 third_party-dev_default

./third_party-dev_default:
total 12
drwxr-xr-x 3 srussell eng 4096 Jul 26 14:25 .
drwxr-xr-x 3 srussell eng 4096 Jul 26 14:25 ..
drwxr-xr-x 2 srussell eng 4096 Jul 26 14:25 315

./third_party-dev_default/315:
total 124424
drwxr-xr-x 2 srussell eng      4096 Jul 26 14:25 .
drwxr-xr-x 3 srussell eng      4096 Jul 26 14:25 ..
-rw------- 1 srussell eng 127399504 Jul 26 14:25 third_party-577247-default-9607b03522d0-linux.tgz
```

The permissions are 600 with the 1.16.0 and later releases. This is also confirmed in the 1.17.1 release. Only workaround is to revert to 1.15.1 or earlier.

**Comment by scottrus** [comment]
This commit https://github.com/jfrog/jfrog-cli-go/commit/913f9c6322e6ebd449d184461e69963ebabdc762#diff-6491780f9a5f1004cb0140253b556efcR348 in the mergeChunks() function creates the file 0600 instead of letting the system umask take precedence. Looks like in some cases directories may get created 0777 as well. (same commit)

**Comment by barbelity** [comment]
@nneul & @scottrus,
Following your issue we looked deeper into our download mechanism.
This happens when our CLI downloads files in 'chunks', as eventually the downloaded file was given initial permissions of 0600 (before umask).

We have commited this fix: https://github.com/jfrog/jfrog-cli-go/commit/6a8343da112496d17fdb83b7791dc0069e918f4d, which changes the permissions of a chunked downloaded file to be 0666 before umask. This new behaviour is similar to our CLI versions prior to 1.16.0, and should solve the described situation.

Your feedback will be most appreciated!
